pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/source/batch/BatchSourceExecutor.java

```diff
@@ -69,14 +69,19 @@
   private String intermediateTopicName;
   private volatile Exception currentError = null;
   private volatile boolean isRunning = false;
-  private ExecutorService discoveryThread = Executors.newSingleThreadExecutor(new DefaultThreadFactory("batch-source-discovery"));
+  private ExecutorService discoveryThread;
 
   @Override
   public void open(Map<String, Object> config, SourceContext sourceContext) throws Exception {
     this.config = config;
     this.sourceContext = sourceContext;
     this.intermediateTopicName = SourceConfigUtils.computeBatchSourceIntermediateTopicName(sourceContext.getTenant(),
       sourceContext.getNamespace(), sourceContext.getSourceName()).toString();
+    this.discoveryThread = Executors.newSingleThreadExecutor(
+      new DefaultThreadFactory(
+        String.format("%s-batch-source-discovery",
+          FunctionCommon.getFullyQualifiedName(
+            sourceContext.getTenant(), sourceContext.getNamespace(), sourceContext.getSourceName()))));
```

Comment:
Maybe unrelated, but it would be better to also use `awaitTermination()` in the shutdown, to ensure all pending tasks are gone before we close the classloader.

Reply:
@merlimat that is what we are doing today:

https://github.com/apache/pulsar/blob/master/pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/source/batch/BatchSourceExecutor.java#L223

Reply:
Gotcha. I missed that